File: src/main/scala/org/joshjoyce/lumivore/sync/SyncStream.scala
package org.joshjoyce.lumivore.sync

import java.nio.file.{Path, Paths}

import org.jetlang.channels.Channel
import org.joshjoyce.lumivore.db.SqliteDatabase
import org.joshjoyce.lumivore.io.{DirectoryPathStream, HashUtils}
import org.joshjoyce.lumivore.util.LumivoreLogging

sealed trait SyncCheckResult

case class Unseen(path: Path) extends SyncCheckResult

case class ContentsChanged(path: Path, oldHash: String, hash: String) extends SyncCheckResult

case object SyncDone extends SyncCheckResult

/**
 * Examines the filesystem for paths that have not been synced.
 */
class SyncStream(database: SqliteDatabase) extends LumivoreLogging {
  private var observers: List[Channel[SyncCheckResult]] = Nil

  def addObserver(c: Channel[SyncCheckResult]) {
    observers = c :: observers
  }

  private def notifyObservers(result: SyncCheckResult) {
    observers.foreach(_.publish(result))
  }

  /**
   * Begin at the root path and traverse recursively,
   * returning any files that don't have sync records
   * or that have changed.
   */
  def check(root: Path) = {
    val validExtensions = database.getExtensions.toSet
    log.info("Loaded extensions: " + validExtensions.mkString(", "))
    log.info("Executing from root: " + root.toString)

    val sha1ByPath = database.getSyncs.groupBy(_._1).mapValues(_.head._2)
    DirectoryPathStream.recurse(root.toFile) {
      path => {
        val pathString = path.toString
        val normPath = Paths.get(pathString)
        val loweredPath = pathString.toLowerCase

        if (validExtensions.isEmpty || validExtensions.exists(loweredPath.endsWith)) {
          val syncOpt = sha1ByPath.get(pathString)

          if (!syncOpt.isDefined) {
            log.info("Unseen|" + pathString)
            val unseen = Unseen(normPath)
            notifyObservers(unseen)
          } else {
            val syncHash = syncOpt.get
            val hash = HashUtils.hashContents(normPath)

            if (hash != syncHash) {
              log.info("Changed|" + pathString)
              val changed = ContentsChanged(normPath, syncHash, hash)
              notifyObservers(changed)
            } else {
              log.info("NoChange|" + pathString)
            }
          }
        }
      }
    }
    
    observers.foreach(_.publish(SyncDone))
  }
}

File: src/main/java/org/joshjoyce/lumivore/sync/SyncStream.java
package org.joshjoyce.lumivore.sync;

import org.apache.log4j.Logger;
import org.jetlang.channels.Channel;
import org.joshjoyce.lumivore.db.SqliteDatabase;
import org.joshjoyce.lumivore.io.DirectoryPathStream;
import org.joshjoyce.lumivore.io.HashUtils;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Examines the filesystem for paths that have not been synced.
 */
public class SyncStream {
    private static final Logger log = Logger.getLogger(SyncStream.class);

    private List<Channel<SyncCheckResult>> observers;
    private SqliteDatabase db;

    SyncStream(SqliteDatabase db) {
        this.db = db;
        this.observers = new ArrayList<>();
    }

    public void addObserver(Channel<SyncCheckResult> c) {
        observers.add(0, c);
    }

    private void notifyObservers(SyncCheckResult result) {
        observers.forEach(c -> c.publish(result));
    }

    /**
     * Begin at the root path and traverse recursively,
     * returning any files that don't have sync records
     * or that have changed.
     */
    public void check(Path root) {
        var validExtensions = db.getExtensions();
        log.info("Loaded extensions: " + String.join(", ", validExtensions));
        log.info("Executing from root: " + root.toString());

        var sha1ByPath = db.getSyncs().stream()
                .collect(Collectors.groupingBy(sync -> sync.path))
                .entrySet().stream().
                        collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().get(0).hash));

        DirectoryPathStream.recurse(root.toFile(), path -> {
            var pathString = path.toString();
            var normPath = Paths.get(pathString);
            var loweredPath = pathString.toLowerCase();

            if (validExtensions.isEmpty() || validExtensions.stream().anyMatch(loweredPath::endsWith)) {
                var syncOpt = sha1ByPath.get(pathString);

                if (syncOpt == null) {
                    log.info("Unseen|" + pathString);
                    notifyObservers(SyncCheckResult.unseen(normPath.toString()));
                } else {
                    var hash = HashUtils.hashContents(normPath);

                    if (!hash.equals(syncOpt)) {
                        log.info("Changed|" + pathString);
                        notifyObservers(SyncCheckResult.contentsChanged(normPath.toString(), syncOpt, hash));
                    } else {
                        log.info("NoChange|" + pathString);
                    }
                }
            }
        });

        observers.forEach(c -> c.publish(SyncCheckResult.done()));
    }
}

